conformance/lang/expressions/raw_template_expr.balt

```diff
@@ -0,0 +1,690 @@
+
+class MyRawTemplate {
+    *object:RawTemplate;
+
+    public final readonly & string[] strings = [];
+    public final readonly & int[] insertions = [];
+}
+
+function init() {
+    int i = 1;
+    MyRawTemplate a = `number ${i} and number ${2}`;
+
+    io:println(a is object:RawTemplate); // @output true
+    io:println(a.strings); // @output ["number "," and number ",""]
+    io:println(a.insertions); // @output [1,2]
+
+    object:RawTemplate b = a;
+    io:println(b is MyRawTemplate); // @output true
+}
+
+Test-Case: panic
+Description: Test the raw template `strings` fields being read-only when required by the contextually-expected type.
+Fail-Issue: ballerina-platform/ballerina-lang#36200
+Labels: array-type, intersection-type, list-constructor-expr, module-class-defn, object:RawTemplate, raw-template-expr
+
+class MyRawTemplate {
+    *object:RawTemplate;
+
+    public final readonly & string[] strings = [];
+    public final readonly & int[] insertions = [];
+}
+
+function init() {
+    int i = 1;
+    MyRawTemplate a = `number ${i} and number ${2}`;
+    object:RawTemplate b = a;
+    b.strings = []; // @panic cannot overwrite the read-only `strings` field
+}
+
+Test-Case: panic
+Description: Test the raw template `insertions` fields being read-only when required by the contextually-expected type.
+Fail-Issue: ballerina-platform/ballerina-lang#36200
+Labels: array-type, intersection-type, list-constructor-expr, module-class-defn, object:RawTemplate, raw-template-expr
+
+class MyRawTemplate {
+    *object:RawTemplate;
+
+    public final readonly & string[] strings = [];
+    public final readonly & int[] insertions = [];
+}
+
+function init() {
+    int i = 1;
+    MyRawTemplate a = `number ${i} and number ${2}`;
+    object:RawTemplate b = a;
+    b.insertions = []; // @panic cannot overwrite the read-only `insertions` field
+}
+
+Test-Case: output
+Description: Test the value of the `strings` field being constructed only once by the class for the raw template expression.
+Fail-Issue: ballerina-platform/ballerina-lang#36354
+Labels: array-type, object:RawTemplate, raw-template-expr
+
+function init() {
+    object:RawTemplate a = fn();
+    string[] b = a.strings;
+    string[] c = fn().strings;
+    io:println(b === c); // @output true
+}
+
+function fn() returns object:RawTemplate {
+    int i = 2;
+    return `foo${1}bar${i}`;
+}
+
+Test-Case: output
+Description: Test the value of the `insertions` field being constructed once for each evaluation of the raw template expression.
+Labels: any, array-type, error-type, object:RawTemplate, raw-template-expr, union-type
+
+function init() {
+    object:RawTemplate a = fn();
+    (any|error)[] b = a.insertions;
+    (any|error)[] c = fn().insertions;
+    io:println(b === c); // @output false
+}
+
+function fn() returns object:RawTemplate {
+    int i = 2;
+    return `foo${1}bar${i}`;
+}
+
```

Comment:
Do we have tests to cover?

> The type of the insertions field is determined from the contextually expected type or the static type of the expressions in the BacktickString in the same way as with a list-constructor-expr.

Reply:
Tests such as https://github.com/ballerina-platform/ballerina-spec/blob/4547bc6db2cdc2ec7de60f5e1bc6319d834f6fa4/conformance/lang/expressions/raw_template_expr.balt#L166 cover the scenario where there is a contextually-expected type IMO.

Will add a test with var for when there's no contextually-expected type.